### DISCUSSION TITLE
azure_sd_config

### DISCUSSION THREAD

**Discussion by shreddy99** [prometheus/prometheus#10723] [Q&A]
Azure service discovery is failing with below error for azure VMSS. user managed identity is already attached to azure vmss. Can someone share some documentation/steps on how to discover azure vmss from on-prem prometheus server. 

prometheus[1740]: level=error ts=2022-05-20T17:13:43.569Z caller=refresh.go:79 component="discovery manager scrape" discovery=azure msg="Unable to refresh target groups" err="could not get virtual machines: could not list virtual machines: compute.VirtualMachinesClient#ListAll: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code=\"AuthorizationFailed\" Message=\"The client '***************' with object id '******************' does not have authorization to perform action 'Microsoft.Compute/virtualMachines/read' over scope '/subscriptions/******************' or the scope is invalid. If access was recently granted, please refresh your credentials.\""

prometheus.yaml contains below

- job_name: azure_service_discovery
  scrape_interval: 1m
  scrape_timeout: 1m
  metrics_path: /metrics
  azure_sd_configs:
    - authentication_method: ManagedIdentity
      environment: AzurePublicCloud
      port: 9273
      subscription_id: ********************
  relabel_configs:
    - source_labels: [__meta_azure_machine_scale_set]
      regex: true.*
      action: keep

**Comment by roidelapluie** [comment]
It seems like a permission issue, not an issue with Prometheus.

**Comment by jeremiahharbach** [comment]
As it says, the client/managed identity/service principal that is performing the service discovery needs to be granted the permission "Microsoft.Compute/virtualMachines/read" at the subscription level within your Azure instance.

For scale sets, you will need to grant the following permissions:
```
Microsoft.Compute/virtualMachineScaleSets/virtualMachines/networkInterfaces/read
Microsoft.Compute/virtualMachineScaleSets/virtualMachines/read
Microsoft.Compute/virtualMachineScaleSets/read               # Not sure if this one is needed or not
Microsoft.Compute/virtualMachines/read
Microsoft.Network/networkInterfaces/read
```
